In [1]:
import os
import sys
import numpy as np
import pylab as pl

%pylab inline

Populating the interactive namespace from numpy and matplotlib


## NULL HYPOTHESIS: the % of former prisoners employed after release is the same or lower for candidates who participated in the program as for the control group, significance level p=0.05
 $H_0: P_0 - P_1 \geq$ 0

 $H_a: P_0 - P_1 $< 0

 $\alpha$ = 0.05

### this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate was or was not ever employed in a CEO transitional job (second row in the table):

 $P_0=0.035, P_1=0.701$

In [2]:
# first thing: choose your significance threshold
alpha=0.05

# we like fractions better then percentages. as a rule of thumb, either use fractions or counts. 
# 0 for control group, 1 for program group

P_0 = 3.5 * 0.01 
P_1 = 70.1 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

#sample sizes 

n_0 = 409
n_1 = 564
    
#lets get the counts by multiplying by the sample size

Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

we must assess the statistical significance


# WHICH TEST?
samples, categorical data

### TWO OPTIONS z test, or chi-square test.

# START WITH Z TEST
#### the z test compares the stanrard deviation of the expected distribution and the observed result. it tells you literally how many standard deviations from the tail an observation is, under the assumption of normality

In [3]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.4210472764645426


In [4]:
# calculate the standard error
# and calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

#### z score: how many standard deviation away from the population parameter is my statistic?

$z=\frac{P_1-P_0}{\sigma}$

In [5]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_2y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_2y)

20.7697865408


In [6]:
## p-value for employment after 2 years: 
## since the largest number we read off the table for is (way) smaller 
## than the value for our statistic 
## our p-value will be smaller than it would be if calculated using 
## (e.g.) .9998 (and in fact using 1.0000 which is the largest number 
## in the table). Using 0.9998 is a **conservative** approach. 

p_2y = 1 - 0.9984


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )
    
report_result(p_2y, alpha)

is the p value 0.00 smaller than the critical value 0.05?
YES!
the Null hypothesis is rejected


# what if we used the values for where the former inmate was or was not "Convicted of a felony" (row 10) in the Recidivism (Years 1-3)?

## NULL HYPOTHESIS: the % of former prisoners reoffend after release is the same or higher for candidates who participated in the program as for the control group, significance level p=0.05
 $H_0: P_0 - P_1 \geq$ 0

 $H_a: P_0 - P_1 $< 0

 $\alpha$ = 0.05

### this is a TEST OF PROPORTIONS. we use the Binomial distribution since it is a yes/no (bernulli) test for each subject: the former inmate reoffended or did not reoffend (tenth row in the table):

 $P_0=0.117, P_1=0.100

In [7]:
# first thing: choose your significance threshold
alpha=0.05

# we like fractions better then percentages. as a rule of thumb, either use fractions or counts. 
# 0 for control group, 1 for program group

P_0 = 11.7 * 0.01 
P_1 = 10.0 * 0.01

if P_0 - P_1 >= 0:
    # we are done
    print ("the Null holds")
else:
    print ("we must assess the statistical significance")

#sample sizes 

n_0 = 409
n_1 = 564
    
#lets get the counts by multiplying by the sample size

Nt_0 = P_0 * n_0
Nt_1 = P_1 * n_1

the Null holds


In [8]:
#define the sample proportion first
sp = (P_0 * n_0 + P_1 * n_1) / (n_1 + n_0)
print (sp)

0.1071459403905447


In [9]:
# calculate the standard error
# and calculate p

p = lambda p0, p1, n0, n1: (p0 * n0 + p1 * n1) / (n0 + n1)
#standard error
se = lambda p, n0, n1: np.sqrt(p * (1 - p) * (1.0 / n0 + 1.0 / n1))

In [10]:
zscore = lambda p0, p1, s : (p0 - p1) / s
z_3y = zscore(P_1, P_0, se(p(P_0, P_1, n_0, n_1), n_0, n_1))
print (z_3y)

-0.846282982605


In [11]:
## p-value for convicted a fellony after 3 years: 
## 0.419

p_3y = 0.419


def report_result(p,a):
    print ('is the p value ' + 
           '{0:.2f} smaller than the critical value {1:.2f}?'.format(p,a))
    if p < a:
        print ("YES!")
    else: 
        print ("NO!")
    
    print ('the Null hypothesis is {}'.format(\
                            'rejected' if p < a  else 'not rejected') )
    
report_result(p_3y, alpha)

is the p value 0.42 smaller than the critical value 0.05?
NO!
the Null hypothesis is not rejected


## Now lets do it with the $\chi^2$ test

| | **success** | **failure** | | 
| :----------:| :----------:  | :------------: | :----------:| 
| test sample	| number of successes in test | number of failures in test | number members of test sample| 
| control sample | number of successes in control | number of failures in control | number members of control sample | 
| | total successes | total failures | number of all members |

 | **employed in subsadized job** | **employed** | **not employed** | | 
| :----------:| :----------:  | :------------: | :----------:| 
| test sample	| $0.701 * 564$ | $0.299 * 564$ | 564| 
| control sample | $0.035 * 409$ | $0.965 * 4.09$ | 409 | 
| total | 409.679 | 562.912 | 973 |

In [12]:
#function for evaluating ChiSquare
def evalChisq(values):
    values = np.array(values)
    E = np.empty_like(values)
    for j in range(len(values[0])):
        for i in range(2):
            
            E[i][j] = ((values[i,:].sum() * values[:,j].sum()) / 
                        (values).sum())
    return ((values - E)**2 / E).sum()

In [13]:
Ntot = 973 # a + b + c + d = tot

sample_values = np.array([[0.701 * 564, 0.299 * 564], [0.0305 * 409, 0.965 * 409]])

print (evalChisq(sample_values))

436.223462575


This number must be compared to the chi sq distribution. You must calculate the number of degrees of freedom forthis experiment. Generally: DOF = Number of observations - number of Independent Variables

so here DOF = 1. Now you can look at the table below and draw conclusions about the rejection of the Null

436 is hella larger then 3.84

why am i mentioning 3.84?

how does the chi square statistic that you derived compare?

please state what that means in terms of your Null hypothesis in a markdown cell below!

### Comment
Our chi square statistics (436)  is well beyond our 0.05 p value (3.84). This means that the Null Hypothesis is rejected confidently.

 | **convicted a fellony** | **yes** | **no** | | 
| :----------:| :----------:  | :------------: | :----------:| 
| test sample	| $0.100 * 568$ | $0.900 * 568$ | 568| 
| control sample | $0.117 * 409$ | $0.883 * 4.09$ | 409 | 
| total | 104.653 | 872.347 | 977 |

In [14]:
Ntot = 977 # a + b + c + d = tot

sample_values = np.array([[0.100 * 568, 0.900 * 568], [0.117 * 409, 0.883 * 409]])

print (evalChisq(sample_values))

0.718493917505


Our chi square statistics (0.72) is between 0.50(0.455) and 0.25(1.32) p-value. So, for a 0.05 significance level, it's not enough to reject Null Hypothesis. 

The results could be due to mere chance.